src/base/http/connection.cpp

```diff
@@ -58,37 +60,60 @@ Connection::~Connection()
 void Connection::read()
 {
     m_idleTimer.restart();
-
     m_receivedData.append(m_socket->readAll());
-    Request request;
-    RequestParser::ErrorCode err = RequestParser::parse(m_receivedData, request);  // TODO: transform request headers to lowercase
-
-    switch (err) {
-    case RequestParser::IncompleteRequest:
-        // Partial request waiting for the rest
-        break;
-
-    case RequestParser::BadRequest:
-        sendResponse(Response(400, "Bad Request"));
-        m_receivedData.clear();
-        break;
-
-    case RequestParser::NoError:
-        const Environment env {m_socket->localAddress(), m_socket->localPort(), m_socket->peerAddress(), m_socket->peerPort()};
-
-        Response response = m_requestHandler->processRequest(request, env);
-        if (acceptsGzipEncoding(request.headers["accept-encoding"]))
-            response.headers[HEADER_CONTENT_ENCODING] = "gzip";
-        sendResponse(response);
-        m_receivedData.clear();
-        break;
+
+    while (!m_receivedData.isEmpty()) {
+        const RequestParser::ParseResult result = RequestParser::parse(m_receivedData);
+
+        switch (result.status) {
+        case RequestParser::ParseStatus::Incomplete: {
+            const long bufferLimit = RequestParser::MAX_REQUEST_SIZE * 1.1;  // some margin for headers
+            if (m_receivedData.size() > bufferLimit) {
+                Logger::instance()->addMessage(tr("Http request size exceeds limiation, closing socket. Limit: %ld, IP: %s")
+                                               .arg(bufferLimit).arg(m_socket->peerAddress().toString()), Log::WARNING);
+                m_socket->close();
```

Comment:
Shouldn't we send response before (e.g. Bad Request)?

Reply:
No, this is `Incomplete` case, means the data we have is incomplete, need to wait until more data comes in and then we try to parse again.

Reply:
>try to parse again.

But you're closing the socket here!..

Reply:
It seems like wrong logic in this method.

Reply:
>But you're closing the socket here!..

It closes **only** when the unparsed data accumulates too much (over `bufferLimit`), otherwise it waits for more data comes in.

Reply:
>It closes only when the unparsed data accumulates too much (over bufferLimit), otherwise it waits for more data comes in.

But why it closes silently, without "Bad Request" response?

Reply:
Now I've found something else.
When request is too long RequestParser returns BadRequest even it is still incomplete so this condition never be true.

Reply:
>But why it closes silently, without "Bad Request" response?

I'll send "413 Payload Too Large".

>When request is too long RequestParser returns BadRequest even it is still incomplete so this condition never be true.

It could be true for a malformed request that has a lot of headers (not included in contentLength) that exceeds this limit and the body has yet arrived.

Reply:
Then I'll repeat my other question. Why don't you apply the MAX_REQUEST_SIZE to the entire request in RequestParser? This would avoid to re-do the check here.
I.e. is request with 19 MB content and 19 MB headers valid for you (when MAX_REQUEST_SIZE is 20 MB)?

Reply:
>Why don't you apply the MAX_REQUEST_SIZE to the entire request in RequestParser? 

you mean like `if (data.size() > MAX_REQUEST_SIZE) return {ParseStatus::BadRequest, Request(), 0};` ?
No, because `data` could have multiple requests concatenated together so by definition that would be wrong.

>I.e. is request with 19 MB content and 19 MB headers valid for you (when MAX_REQUEST_SIZE is 20 MB)?

It should be valid, but the header size is unrealistic so if it got dropped is also acceptable IMO.

Reply:
>you mean like if (data.size() > MAX_REQUEST_SIZE) return {ParseStatus::BadRequest, Request(), 0}; ?
No, because data could have multiple requests concatenated together so by definition that would be wrong.

No. I mean `if ((headerSize + contentLength) > MAX_REQUEST_SIZE) return {ParseStatus::BadRequest, Request(), 0};`

Reply:
>No. I mean if ((headerSize + contentLength) > MAX_REQUEST_SIZE) return {ParseStatus::BadRequest, Request(), 0};

Let me say it in another way, as I said, we should limit the content length directly because we want to control the POST payload size.
If we limit by (headerSize + contentLength), I don't like this because there is no sensible limit on how large the header can be and from recent trends, header gets larger than ever.

Reply:
> I don't like this because there is no sensible limit on how large the header can be

But you still limit it by 10% from content. Just append it to MAX_REQUEST_SIZE.
>header gets larger than ever

Why?
In my case it can be at most MAX_REQUEST_SIZE. In your case it can be at most 1.1*MAX_REQUEST_SIZE.

Reply:
Well... It seems that if we do not introduce a separate limit for the size of headers, the remaining ways are reduced to the same. So do whatever you want. But I still handled it in one place (in RequestParser).

Reply:
>So do whatever you want. But I still handled it in one place (in RequestParser).

I'll keep it for now, at least I want to make sure `m_receivedData` size won't explode.